File: backend/analyzer/log_parser.py
"""Log file parsing utilities supporting multiple formats."""

import re
from datetime import datetime
from typing import List, Dict, Tuple, Optional
from dataclasses import dataclass


@dataclass
class ParsedLogEntry:
    """Represents a parsed log entry."""
    ip: str
    timestamp: Optional[datetime]
    method: str
    path: str
    protocol: str
    status: int
    size: int
    raw_line: str


# Regex patterns for different log formats
# Apache Combined Log Format with optional referrer, user-agent, and response time
APACHE_COMBINED_REGEX = re.compile(
    r'^(\S+)\s+\S+\s+\S+\s+\[([^\]]+)\]\s+"(\S+)\s+(\S+)\s*(\S*)"\s+(\d+)\s+(\S+)(?:\s+"([^"]*)"\s+"([^"]*)")?(?:\s+(\d+))?'
)

NGINX_REGEX = re.compile(
    r'^(\S+)\s+-\s+-\s+\[([^\]]+)\]\s+"(\S+)\s+(\S+)\s*(\S*)"\s+(\d+)\s+(\S+)(?:\s+"([^"]*)"\s+"([^"]*)")?(?:\s+(\d+))?'
)

# Month mapping for Apache date format
MONTH_MAP = {
    'Jan': 1, 'Feb': 2, 'Mar': 3, 'Apr': 4, 'May': 5, 'Jun': 6,
    'Jul': 7, 'Aug': 8, 'Sep': 9, 'Oct': 10, 'Nov': 11, 'Dec': 12
}

# CSV header keywords for auto-detection
CSV_HEADER_KEYWORDS = ['ip', 'address', 'timestamp', 'date', 'method', 'url', 'path', 'status', 'size', 'bytes']


def parse_apache_timestamp(timestamp_str: str) -> Optional[datetime]:
    """Parse Apache log timestamp format: 15/Jul/2009:14:58:59 -0700"""
    try:
        match = re.match(r'(\d+)/(\w+)/(\d+):(\d+):(\d+):(\d+)\s*([+-]\d+)?', timestamp_str)
        if match:
            day, month, year, hour, minute, second, tz = match.groups()
            month_num = MONTH_MAP.get(month, 1)
            return datetime(
                int(year), month_num, int(day),
                int(hour), int(minute), int(second)
            )
    except (ValueError, TypeError):
        pass
    return None


def parse_log_line(line: str) -> Optional[ParsedLogEntry]:
    """Parse a single log line using multiple format patterns."""
    line = line.strip()
    if not line:
        return None
    
    # Try Apache/Nginx combined format
    for pattern in [APACHE_COMBINED_REGEX, NGINX_REGEX]:
        match = pattern.match(line)
        if match:
            ip, timestamp_str, method, path, protocol, status, size = match.groups()
            
            # Parse size, handling '-' for missing values
            parsed_size = 0 if size == '-' else int(size) if size.isdigit() else 0
            
            # Skip entries with size '-' (304 responses, etc.) as per requirements
            if size == '-':
                return None
            
            return ParsedLogEntry(
                ip=ip,
                timestamp=parse_apache_timestamp(timestamp_str),
                method=method or 'GET',
                path=path or '/',
                protocol=protocol or 'HTTP/1.1',
                status=int(status) if status.isdigit() else 0,
                size=parsed_size,
                raw_line=line
            )
    
    return None


def detect_csv_format(first_line: str) -> bool:
    """Detect if the file is in CSV format."""
    lower_line = first_line.lower()
    return any(keyword in lower_line for keyword in CSV_HEADER_KEYWORDS) or ',' in first_line


def parse_csv_line(values: List[str], field_map: Dict[str, int], line: str) -> Optional[ParsedLogEntry]:
    """Parse a CSV line based on detected field mapping."""
    try:
        ip = values[field_map.get('ip', -1)] if field_map.get('ip', -1) >= 0 else 'unknown'
        
        timestamp = None
        if field_map.get('timestamp', -1) >= 0:
            ts_str = values[field_map['timestamp']]
            try:
                timestamp = datetime.fromisoformat(ts_str.replace('Z', '+00:00'))
            except ValueError:
                timestamp = parse_apache_timestamp(ts_str)
        
        method = values[field_map.get('method', -1)] if field_map.get('method', -1) >= 0 else 'GET'
        path = values[field_map.get('path', -1)] if field_map.get('path', -1) >= 0 else '/'
        
        status = 0
        if field_map.get('status', -1) >= 0:
            status_str = values[field_map['status']]
            status = int(status_str) if status_str.isdigit() else 0
        
        size = 0
        if field_map.get('size', -1) >= 0:
            size_str = values[field_map['size']]
            if size_str and size_str != '-':
                size = int(size_str) if size_str.isdigit() else 0
            else:
                return None  # Skip entries with no size
        
        # Skip entries with insufficient data
        if ip == 'unknown' and path == '/' and status == 0:
            return None
        
        return ParsedLogEntry(
            ip=ip,
            timestamp=timestamp,
            method=method,
            path=path,
            protocol='HTTP/1.1',
            status=status,
            size=size,
            raw_line=line
        )
    except (IndexError, ValueError):
        return None


def detect_csv_fields(headers: List[str]) -> Dict[str, int]:
    """Map CSV headers to field indices."""
    field_map = {}
    
    for i, header in enumerate(headers):
        h = header.lower().strip().strip('"\'')
        
        if 'ip' in h or 'address' in h or 'client' in h:
            field_map['ip'] = i
        elif 'time' in h or 'date' in h:
            field_map['timestamp'] = i
        elif 'method' in h or 'verb' in h:
            field_map['method'] = i
        elif 'path' in h or 'url' in h or 'uri' in h or 'request' in h:
            field_map['path'] = i
        elif 'status' in h or 'code' in h or 'response' in h:
            field_map['status'] = i
        elif 'size' in h or 'bytes' in h or 'length' in h:
            field_map['size'] = i
    
    return field_map


def parse_csv_file(lines: List[str]) -> Tuple[List[ParsedLogEntry], List[str]]:
    """Parse a CSV file."""
    entries = []
    errors = []
    
    if not lines:
        return entries, ['Empty file']
    
    # Detect delimiter
    first_line = lines[0]
    delimiter = '\t' if '\t' in first_line else ','
    
    # Parse header
    headers = [h.strip() for h in first_line.split(delimiter)]
    field_map = detect_csv_fields(headers)
    
    if 'ip' not in field_map and 'path' not in field_map:
        errors.append('CSV does not contain recognizable log fields (ip, path, status, etc.)')
        return entries, errors
    
    # Parse data rows
    for i, line in enumerate(lines[1:], start=2):
        line = line.strip()
        if not line:
            continue
        
        values = [v.strip().strip('"\'') for v in line.split(delimiter)]
        entry = parse_csv_line(values, field_map, line)
        
        if entry:
            entries.append(entry)
        else:
            errors.append(f'Line {i}: Unable to parse - "{line[:60]}..."')
    
    return entries, errors


def parse_log_file(content: str) -> Tuple[List[ParsedLogEntry], List[str]]:
    """
    Parse log file content and return entries and errors.
    Supports Apache, Nginx, and CSV formats.
    """
    lines = [line for line in content.split('\n') if line.strip()]
    entries = []
    errors = []
    
    if not lines:
        return entries, ['Empty file']
    
    # Detect format
    if detect_csv_format(lines[0]):
        return parse_csv_file(lines)
    
    # Parse as standard log format
    for i, line in enumerate(lines, start=1):
        entry = parse_log_line(line)
        if entry:
            entries.append(entry)
        else:
            errors.append(f'Line {i}: Unable to parse - "{line[:60]}..."')
    
    return entries, errors

File: lib/log-parser.ts
import type { ParsedLogEntry } from "./types"

// Apache Combined Log Format regex (with optional referrer, user-agent, response-time)
// Format: IP - - [timestamp] "method path protocol" status size "referrer" "user-agent" [response-time]
const APACHE_COMBINED_REGEX = /^(\S+)\s+\S+\s+\S+\s+\[([^\]]+)\]\s+"(\S+)\s+(\S+)\s*(\S*)"\s+(\d+)\s+(\S+)(?:\s+"([^"]*)"\s+"([^"]*)")?(?:\s+(\d+))?/

// Nginx default format regex (with optional referrer, user-agent, response-time)
const NGINX_REGEX = /^(\S+)\s+-\s+-\s+\[([^\]]+)\]\s+"(\S+)\s+(\S+)\s*(\S*)"\s+(\d+)\s+(\S+)(?:\s+"([^"]*)"\s+"([^"]*)")?(?:\s+(\d+))?/

// Common Log Format regex (basic format without referrer/user-agent)
const COMMON_LOG_REGEX = /^(\S+)\s+\S+\s+\S+\s+\[([^\]]+)\]\s+"(\S+)\s+(\S+)\s*(\S*)"\s+(\d+)\s+(\S+)/

// CSV format detection
const CSV_HEADER_KEYWORDS = ["ip", "address", "timestamp", "date", "method", "url", "path", "status", "size", "bytes"]

export async function parseLogFileAsync(
  content: string,
  onProgress?: (progress: number, message: string) => void,
): Promise<{ entries: ParsedLogEntry[]; errors: string[] }> {
  const lines = content.split("\n").filter((line) => line.trim())
  const entries: ParsedLogEntry[] = []
  const errors: string[] = []

  if (lines.length === 0) {
    return { entries, errors: ["Empty file"] }
  }

  // Detect format - prioritize log format detection over CSV
  // Check if first line looks like a log entry (has timestamp in brackets, IP, etc.)
  const firstLine = lines[0]
  const looksLikeLog = /^\S+\s+-\s+-\s+\[/.test(firstLine) || /^\S+\s+\S+\s+\S+\s+\[/.test(firstLine)
  
  // Only detect as CSV if it doesn't look like a log AND has CSV characteristics
  const firstLineLower = firstLine.toLowerCase()
  const hasHeaderKeywords = CSV_HEADER_KEYWORDS.some((keyword) => firstLineLower.includes(keyword))
  const isCSV = !looksLikeLog && hasHeaderKeywords && firstLine.includes(",") && !firstLine.match(/\[\d{1,2}\/\w{3}\/\d{4}/)

  if (isCSV) {
    return parseCSVAsync(lines, onProgress)
  }

  // Process in chunks to prevent UI freeze
  const CHUNK_SIZE = 10000
  const totalLines = lines.length

  for (let i = 0; i < totalLines; i += CHUNK_SIZE) {
    const end = Math.min(i + CHUNK_SIZE, totalLines)

    for (let j = i; j < end; j++) {
      const line = lines[j].trim()
      if (!line) continue

      const parsed = parseLogLine(line)
      if (parsed) {
        entries.push(parsed)
      } else {
        errors.push(`Line ${j + 1}: Unable to parse - "${line.substring(0, 80)}..."`)
      }
    }

    // Report progress and yield to main thread
    const progress = Math.round((end / totalLines) * 100)
    onProgress?.(progress, `Parsing log entries... ${progress}% (${entries.length.toLocaleString()} found)`)

    // Yield to main thread every chunk
    await new Promise((resolve) => setTimeout(resolve, 0))
  }

  return { entries, errors }
}

export function parseLogFile(content: string): { entries: ParsedLogEntry[]; errors: string[] } {
  const lines = content.split("\n").filter((line) => line.trim())
  const entries: ParsedLogEntry[] = []
  const errors: string[] = []

  if (lines.length === 0) {
    return { entries, errors: ["Empty file"] }
  }

  // Detect format - prioritize log format detection over CSV
  const firstLine = lines[0]
  const looksLikeLog = /^\S+\s+-\s+-\s+\[/.test(firstLine) || /^\S+\s+\S+\s+\S+\s+\[/.test(firstLine)
  
  // Only detect as CSV if it doesn't look like a log AND has CSV characteristics
  const firstLineLower = firstLine.toLowerCase()
  const hasHeaderKeywords = CSV_HEADER_KEYWORDS.some((keyword) => firstLineLower.includes(keyword))
  const isCSV = !looksLikeLog && hasHeaderKeywords && firstLine.includes(",") && !firstLine.match(/\[\d{1,2}\/\w{3}\/\d{4}/)

  if (isCSV) {
    return parseCSV(lines)
  }

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i].trim()
    if (!line) continue

    const parsed = parseLogLine(line)
    if (parsed) {
      entries.push(parsed)
    } else {
      errors.push(`Line ${i + 1}: Unable to parse - "${line.substring(0, 80)}..."`)
    }
  }

  return { entries, errors }
}

function parseLogLine(line: string): ParsedLogEntry | null {
  // Try Apache Combined format first (with referrer, user-agent, response-time)
  let match = line.match(APACHE_COMBINED_REGEX)
  if (match) {
    const [, ip, timestampStr, method, path, protocol, statusStr, sizeStr] = match
    return {
      ip,
      timestamp: parseTimestamp(timestampStr),
      method: method || "GET",
      path: path || "/",
      protocol: protocol || "HTTP/1.1",
      status: Number.parseInt(statusStr) || 0,
      size: sizeStr === "-" ? 0 : Number.parseInt(sizeStr) || 0,
      rawLine: line,
    }
  }

  // Try Nginx format
  match = line.match(NGINX_REGEX)
  if (match) {
    const [, ip, timestampStr, method, path, protocol, statusStr, sizeStr] = match
    return {
      ip,
      timestamp: parseTimestamp(timestampStr),
      method: method || "GET",
      path: path || "/",
      protocol: protocol || "HTTP/1.1",
      status: Number.parseInt(statusStr) || 0,
      size: sizeStr === "-" ? 0 : Number.parseInt(sizeStr) || 0,
      rawLine: line,
    }
  }

  // Try Common Log Format (basic)
  match = line.match(COMMON_LOG_REGEX)
  if (match) {
    const [, ip, timestampStr, method, path, protocol, statusStr, sizeStr] = match
    return {
      ip,
      timestamp: parseTimestamp(timestampStr),
      method: method || "GET",
      path: path || "/",
      protocol: protocol || "HTTP/1.1",
      status: Number.parseInt(statusStr) || 0,
      size: sizeStr === "-" ? 0 : Number.parseInt(sizeStr) || 0,
      rawLine: line,
    }
  }

  return null
}

function parseTimestamp(str: string): string | null {
  if (!str || str === "-") return null

  try {
    // Format: 15/Jul/2009:14:58:59 -0700
    const apacheMatch = str.match(/(\d{1,2})\/(\w{3})\/(\d{4}):(\d{2}):(\d{2}):(\d{2})\s*([+-]\d{4})?/)
    if (apacheMatch) {
      const [, day, month, year, hour, minute, second] = apacheMatch
      const monthMap: Record<string, number> = {
        Jan: 0,
        Feb: 1,
        Mar: 2,
        Apr: 3,
        May: 4,
        Jun: 5,
        Jul: 6,
        Aug: 7,
        Sep: 8,
        Oct: 9,
        Nov: 10,
        Dec: 11,
      }
      const monthNum = monthMap[month]
      if (monthNum !== undefined) {
        const date = new Date(
          Number.parseInt(year),
          monthNum,
          Number.parseInt(day),
          Number.parseInt(hour),
          Number.parseInt(minute),
          Number.parseInt(second),
        )
        if (!isNaN(date.getTime())) {
          return date.toISOString()
        }
      }
    }

    if (str.includes("T") || str.includes("-")) {
      const date = new Date(str)
      if (!isNaN(date.getTime())) {
        return date.toISOString()
      }
    }

    // Format: 2009-07-15 14:58:59
    const isoLikeMatch = str.match(/(\d{4})-(\d{2})-(\d{2})\s+(\d{2}):(\d{2}):(\d{2})/)
    if (isoLikeMatch) {
      const [, year, month, day, hour, minute, second] = isoLikeMatch
      const date = new Date(
        Number.parseInt(year),
        Number.parseInt(month) - 1,
        Number.parseInt(day),
        Number.parseInt(hour),
        Number.parseInt(minute),
        Number.parseInt(second),
      )
      if (!isNaN(date.getTime())) {
        return date.toISOString()
      }
    }

    const usDateMatch = str.match(/(\d{1,2})\/(\d{1,2})\/(\d{4})\s+(\d{2}):(\d{2}):(\d{2})/)
    if (usDateMatch) {
      const [, month, day, year, hour, minute, second] = usDateMatch
      const date = new Date(
        Number.parseInt(year),
        Number.parseInt(month) - 1,
        Number.parseInt(day),
        Number.parseInt(hour),
        Number.parseInt(minute),
        Number.parseInt(second),
      )
      if (!isNaN(date.getTime())) {
        return date.toISOString()
      }
    }

    // Last resort: try native Date parsing
    const date = new Date(str)
    if (!isNaN(date.getTime())) {
      return date.toISOString()
    }
  } catch {
    // Ignore parsing errors
  }
  return null
}

async function parseCSVAsync(
  lines: string[],
  onProgress?: (progress: number, message: string) => void,
): Promise<{ entries: ParsedLogEntry[]; errors: string[] }> {
  const entries: ParsedLogEntry[] = []
  const errors: string[] = []

  const firstLine = lines[0]
  const delimiter = firstLine.includes("\t") ? "\t" : ","
  const headers = firstLine.split(delimiter).map((h) => h.trim().toLowerCase().replace(/['"]/g, ""))

  const fieldMap = {
    ip: headers.findIndex((h) => h.includes("ip") || h.includes("address") || h.includes("client")),
    timestamp: headers.findIndex((h) => h.includes("time") || h.includes("date")),
    method: headers.findIndex((h) => h.includes("method") || h.includes("verb")),
    path: headers.findIndex(
      (h) => h.includes("path") || h.includes("url") || h.includes("uri") || h.includes("request"),
    ),
    status: headers.findIndex((h) => h.includes("status") || h.includes("code") || h.includes("response")),
    size: headers.findIndex((h) => h.includes("size") || h.includes("bytes") || h.includes("length")),
  }

  if (fieldMap.ip === -1 && fieldMap.path === -1) {
    errors.push("CSV does not contain recognizable log fields (ip, path, status, etc.)")
    return { entries, errors }
  }

  const CHUNK_SIZE = 10000
  const totalLines = lines.length - 1

  for (let i = 1; i < lines.length; i += CHUNK_SIZE) {
    const end = Math.min(i + CHUNK_SIZE, lines.length)

    for (let j = i; j < end; j++) {
      const line = lines[j].trim()
      if (!line) continue

      try {
        const values = parseCSVLine(line, delimiter)

        const entry: ParsedLogEntry = {
          ip: fieldMap.ip >= 0 ? values[fieldMap.ip] || "unknown" : "unknown",
          timestamp: fieldMap.timestamp >= 0 ? parseTimestamp(values[fieldMap.timestamp]) : null,
          method: fieldMap.method >= 0 ? values[fieldMap.method] || "GET" : "GET",
          path: fieldMap.path >= 0 ? values[fieldMap.path] || "/" : "/",
          protocol: "HTTP/1.1",
          status: fieldMap.status >= 0 ? Number.parseInt(values[fieldMap.status]) || 0 : 0,
          size: fieldMap.size >= 0 ? Number.parseInt(values[fieldMap.size]) || 0 : 0,
          rawLine: line,
        }

        if (entry.ip === "unknown" && entry.path === "/" && entry.status === 0) {
          errors.push(`Line ${j + 1}: Insufficient data - "${line.substring(0, 60)}..."`)
          continue
        }

        entries.push(entry)
      } catch {
        errors.push(`Line ${j + 1}: Parse error - "${line.substring(0, 60)}..."`)
      }
    }

    const progress = Math.round(((end - 1) / totalLines) * 100)
    onProgress?.(progress, `Parsing CSV... ${progress}% (${entries.length.toLocaleString()} found)`)
    await new Promise((resolve) => setTimeout(resolve, 0))
  }

  return { entries, errors }
}

function parseCSV(lines: string[]): { entries: ParsedLogEntry[]; errors: string[] } {
  const entries: ParsedLogEntry[] = []
  const errors: string[] = []

  const firstLine = lines[0]
  const delimiter = firstLine.includes("\t") ? "\t" : ","
  const headers = firstLine.split(delimiter).map((h) => h.trim().toLowerCase().replace(/['"]/g, ""))

  const fieldMap = {
    ip: headers.findIndex((h) => h.includes("ip") || h.includes("address") || h.includes("client")),
    timestamp: headers.findIndex((h) => h.includes("time") || h.includes("date")),
    method: headers.findIndex((h) => h.includes("method") || h.includes("verb")),
    path: headers.findIndex(
      (h) => h.includes("path") || h.includes("url") || h.includes("uri") || h.includes("request"),
    ),
    status: headers.findIndex((h) => h.includes("status") || h.includes("code") || h.includes("response")),
    size: headers.findIndex((h) => h.includes("size") || h.includes("bytes") || h.includes("length")),
  }

  if (fieldMap.ip === -1 && fieldMap.path === -1) {
    errors.push("CSV does not contain recognizable log fields (ip, path, status, etc.)")
    return { entries, errors }
  }

  for (let i = 1; i < lines.length; i++) {
    const line = lines[i].trim()
    if (!line) continue

    try {
      const values = parseCSVLine(line, delimiter)

      const entry: ParsedLogEntry = {
        ip: fieldMap.ip >= 0 ? values[fieldMap.ip] || "unknown" : "unknown",
        timestamp: fieldMap.timestamp >= 0 ? parseTimestamp(values[fieldMap.timestamp]) : null,
        method: fieldMap.method >= 0 ? values[fieldMap.method] || "GET" : "GET",
        path: fieldMap.path >= 0 ? values[fieldMap.path] || "/" : "/",
        protocol: "HTTP/1.1",
        status: fieldMap.status >= 0 ? Number.parseInt(values[fieldMap.status]) || 0 : 0,
        size: fieldMap.size >= 0 ? Number.parseInt(values[fieldMap.size]) || 0 : 0,
        rawLine: line,
      }

      if (entry.ip === "unknown" && entry.path === "/" && entry.status === 0) {
        errors.push(`Line ${i + 1}: Insufficient data - "${line.substring(0, 60)}..."`)
        continue
      }

      entries.push(entry)
    } catch {
      errors.push(`Line ${i + 1}: Parse error - "${line.substring(0, 60)}..."`)
    }
  }

  return { entries, errors }
}

function parseCSVLine(line: string, delimiter: string): string[] {
  const values: string[] = []
  let current = ""
  let inQuotes = false

  for (let i = 0; i < line.length; i++) {
    const char = line[i]

    if (char === '"' && !inQuotes) {
      inQuotes = true
    } else if (char === '"' && inQuotes) {
      if (line[i + 1] === '"') {
        current += '"'
        i++
      } else {
        inQuotes = false
      }
    } else if (char === delimiter && !inQuotes) {
      values.push(current.trim())
      current = ""
    } else {
      current += char
    }
  }
  values.push(current.trim())

  return values
}
